### PR TITLE
Replace type switches in Rollback with Visitor pattern

### PR DESCRIPTION
**Pull request by mahdix**
Fixes https://github.com/kubernetes/kubectl/issues/127

A refactoring to make Rollback module less dependent on internal packages.

### CODE REVIEW

**Comment by mengqiy** [comment]
@mahdix Please add release note according to https://github.com/kubernetes/kubernetes/pull/55557#issuecomment-343752875.

File: pkg/kubectl/rollback.go [hunk 1]
```diff
@@ -56,16 +56,46 @@ type Rollbacker interface {
...
+	v.result = &StatefulSetRollbacker{v.clientset}
+}
+
+func (v *RollbackVisitor) VisitDaemonSet(kind kapps.GroupKindElement) {
+	v.result = &DaemonSetRollbacker{v.clientset}
+}
+
+func (v *RollbackVisitor) VisitJob(kind kapps.GroupKindElement)                   {}
+func (v *RollbackVisitor) VisitPod(kind kapps.GroupKindElement)                   {}
+func (v *RollbackVisitor) VisitReplicaSet(kind kapps.GroupKindElement)            {}
```
Comment (mengqiy) [inline pkg/kubectl/rollback.go]: @janetkuo Should we do this for RS and RC?

File: pkg/kubectl/rollback.go [hunk 2]
```diff
@@ -56,16 +56,46 @@ type Rollbacker interface {
...
+func (v *RollbackVisitor) VisitStatefulSet(kind kapps.GroupKindElement) {
+	v.result = &StatefulSetRollbacker{v.clientset}
+}
+
+func (v *RollbackVisitor) VisitDaemonSet(kind kapps.GroupKindElement) {
+	v.result = &DaemonSetRollbacker{v.clientset}
+}
+
+func (v *RollbackVisitor) VisitJob(kind kapps.GroupKindElement)                   {}
+func (v *RollbackVisitor) VisitPod(kind kapps.GroupKindElement)                   {}
```
[the remark this answers is not in the document]
Comment (janetkuo) [reply pkg/kubectl/rollback.go]: This is the downside of using visitor pattern, as I mentioned in
https://github.com/kubernetes/kubernetes/pull/55257#discussion_r149764557 -- visitors are forced to implement all the `VisitXxxx` functions, even for the resources the visitor doesn't/won't support. `VisitPod` here is just a dummy function for that purpose

File: pkg/kubectl/rollback.go (hunk 1 above)
Comment (janetkuo) [reply pkg/kubectl/rollback.go]: As mentioned above, even though RC/RS don't support rollback, we can't avoid implementing visit functions for them

File: pkg/kubectl/rollback.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mengqiy) [reply pkg/kubectl/rollback.go]: Make sense. Thanks for recalling this.

[the remark this answers is not in the document]
Comment (mengqiy) [reply pkg/kubectl/rollback.go]: cc: @pwittrock

**Comment by mahdix** [comment]
Can someone review and approve this?

**Comment by mahdix** [comment]
ping @janetkuo

**Review by janetkuo — APPROVED** [review]

**Pull request merged**